sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamProgress.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import scala.collection.mutable
+
+/**
+ * Tracks the progress of processing data from one or more [[Source]]s that are present in a
+ * streaming query. This is similar to simplified, single-instance vector clock that progresses
+ * monotonically forward.
+ *
+ * TODO: serialization API
```

Comment:
While that is a reasonable solution, I'm not sure thats exactly how we want to do it.  Here are the requirements as I see them:
 - User created Offsets should be serializable
 - A collection of offsets needs to be serializable in a way that we can reassociate them with their sources on deserialization
 - There should be some class (probably StreamProgress) that acts as a container and makes this opaque to Sinks